unstructured/ingest/connector/astra.py

```diff
@@ -0,0 +1,116 @@
+from unstructured.ingest.error import DestinationConnectionError, SourceConnectionNetworkError
+from unstructured.ingest.interfaces import (
+    AccessConfig,
+    BaseConnectorConfig,
+    BaseDestinationConnector,
+    WriteConfig,
+)
+from unstructured.ingest.logger import logger
+from unstructured.ingest.utils.data_prep import chunk_generator
+from unstructured.staging.base import flatten_dict
+from unstructured.utils import requires_dependencies
+
+
+if t.TYPE_CHECKING:
+    from astrapy.db import AstraDB, AstraDBCollection
+
+NON_INDEXED_FIELDS = ["metadata._node_content", "content"]
+
+
+@dataclass
+class AstraDBAccessConfig(AccessConfig):
+    token: t.Optional[str] = enhanced_field(default=None, sensitive=True)
+    api_endpoint: t.Optional[str] = enhanced_field(default=None, sensitive=True)
+
+
+@dataclass
+class SimpleAstraDBConfig(BaseConnectorConfig):
+    access_config: AstraDBAccessConfig
+    collection_name: str
+    embedding_dimension: int
+
+
+@dataclass
+class AstraDBWriteConfig(WriteConfig):
+    batch_size: int = 20
+
+
+@dataclass
+class AstraDBDestinationConnector(BaseDestinationConnector):
+    write_config: AstraDBWriteConfig
+    connector_config: SimpleAstraDBConfig
+    _astra_db: t.Optional["AstraDB"] = field(init=False, default=None)
+    _astra_db_collection: t.Optional["AstraDBCollection"] = field(init=False, default=None)
+
+    def to_dict(self, **kwargs):
+        """
+        The _astra_db_collection variable in this dataclass breaks deepcopy due to:
+        TypeError: cannot pickle '_thread.lock' object
+        When serializing, remove it, meaning client data will need to be reinitialized
+        when deserialized
+        """
+        self_cp = copy.copy(self)
+
+        if hasattr(self_cp, "_astra_db_collection"):
+            setattr(self_cp, "_astra_db_collection", None)
+        
+        return _asdict(self_cp, **kwargs)
+
+    @property
+    @requires_dependencies(["astrapy"], extras="astra")
+    def astra_db_collection(self) -> "AstraDBCollection":
+        if self._astra_db_collection is None:
+            from astrapy.db import AstraDB
+
+            # Build the Astra DB object
+            self._astra_db = AstraDB(
+                api_endpoint=self.connector_config.access_config.api_endpoint,
+                token=self.connector_config.access_config.token,
+            )
+
+            # Create and connect to the newly created collection
+            self._astra_db_collection = self._astra_db.create_collection(
+                collection_name=self.connector_config.collection_name,
+                dimension=self.connector_config.embedding_dimension,
+                options={"indexing": {"deny": NON_INDEXED_FIELDS}},
+            )
+        return self._astra_db_collection
+
+    @requires_dependencies(["astrapy"], extras="astra")
+    @DestinationConnectionError.wrap
+    def initialize(self):
+        astra_db_collection = self._astra_db_collection
```

Comment:
Should be `_ = self.astra_db_collection`
I should have been more clear. I can see the misunderstanding.

Reply:
My mistake! so basically all references in that file should be `self.astra_db_collection` - no leading underscore / internal method? I assume it's okay to keep `self._astra_db` right, since this shouldn't need to be interacted with directly?

Reply:
Just change it in the 2 spots I suggested. Everything works. Don't change anything else :)
I'll give it a tiny little more checkover when I take it over.
In the other spots, the dataclass attribute shoud stay  _astra_db_collection)

Reply:
And also leave self._astra_db as it is.

Reply:
Heh, got it, yep, i see now the property has the right name. I totally missed that. Just pushed the update! @potter-potter

Reply:
calling self.astra_db_collection will then populate those dataclass attributes. _astra_db and _astra_db_collection